src/main/java/software/amazon/awssdk/crt/http/HttpConnection.java

```diff
@@ -0,0 +1,218 @@
+
+        this.uri = uri;
+        this.clientBootstrap = bootstrap;
+        this.socketOptions = socketOptions;
+        this.tlsContext = tlsContext;
+        this.userConnectionCallbacks = callbacks;
+    }
+
+    /**
+     * Connects to the Http Endpoint
+     * @return Future indicating when
+     */
+    public synchronized CompletableFuture<Boolean> connect() throws CrtRuntimeException {
+        if (!isNull()) {
+            return connectedFuture;
+        }
+
+        String endpoint = uri.getHost();
+        int port = uri.getPort();
+
+        /* Pick a default port if one wasn't set in the URI */
+        if (port == -1) {
+            port = HTTP.equals(uri.getScheme()) ? 80 : 443;
+        }
+
+        connectedFuture = new CompletableFuture<>();
+        connectAck = AsyncCallback.wrapFuture(connectedFuture, false);
+
+        acquire(httpConnectionNew(this,
+                clientBootstrap.native_ptr(),
+                socketOptions.native_ptr(),
+                tlsContext != null ? tlsContext.native_ptr() : 0,
+                endpoint,
+                (short) port));
+
+        connectionState = ConnectionState.CONNECTING;
+
+        return connectedFuture;
+    }
+
+    /**
+     * Disconnects if necessary, and frees native resources associated with this connection
+     */
+    @Override
+    public synchronized void close() {
+        if (!isNull()) {
+            httpConnectionRelease(release());
+            this.connectionState = ConnectionState.DISCONNECTED;
+        }
+
+        super.close();
+
+    }
+
+    /**
+     * Returns the current connection state. This function should not be used often, it is much better to respond
+     * to events delivered via the HttpConnectionEvents interface provided at construction.
+     * @return The current connection state
+     */
+    public synchronized ConnectionState getState() {
+        return connectionState;
+    }
+
+    /** Called from native when the connection is established the first time **/
+    private synchronized void onConnectionComplete(int errorCode) {
+        if (errorCode == AWS_CRT_SUCCESS) {
+            connectionState = ConnectionState.CONNECTED;
+            if (connectAck != null) {
+                connectAck.onSuccess(true);
+                connectAck = null;
+            }
+        } else {
+            connectionState = ConnectionState.DISCONNECTED;
+            if (connectAck != null) {
+                connectAck.onFailure(new HttpException(errorCode));
+                connectAck = null;
+            }
+        }
+        if (userConnectionCallbacks != null) {
+            userConnectionCallbacks.onConnectionComplete(errorCode);
+        }
+    }
+
+    /** Called from native when the connection is shutdown. **/
+    private synchronized void onConnectionShutdown(int errorCode) {
```

Comment:
It appears the only state you're protecting with synchronized is the connectionState. Why not just make that atomic? the underlying c code is thread safe.

Reply:
There's potential for race conditions between the EventLoop Thread and the User Thread if `synchronized` is removed.

Here's an example from the `HttpConnection.connect()` method:
```
// The httpConnectionNew call will schedule a task on the EventLoop to call onConnectionComplete() when the connection is completed (and set the connection's internal state to CONNECTED)
acquire(httpConnectionNew(this,
                clientBootstrap.native_ptr(),
                socketOptions.native_ptr(),
                tlsContext != null ? tlsContext.native_ptr() : 0,
                endpoint,
                (short) port));
// If this method was not synchronized, and the EventLoop task on the above line was completed "instantaneously" here, the User Thread would then overwrite the state from CONNECTED to CONNECTING.
connectionState = ConnectionState.CONNECTING;
```

Even if I migrate to `AtomicReference<ConnectionState>` and do compare-and-swap everywhere, it still opens the door for compare-and-swap failures if the EventLoop is "too fast" or UserThread "too slow", which would be extremely rare and hard to reproduce connection failures.

Reply:
As noted below, the most likely solution is not to track this state at all, since callbacks are available for every state change.

Reply:
Removed `ConnectionState` Enum